### ISSUE TITLE
Snapshot Update While Recording

### ISSUE THREAD

**Issue by jwillaz** [twrecked/hass-aarlo#29]
First of all, thank you Steve for this awesome component! My overall automation responsiveness when working with Arlo cameras in HASS is drastically improved, and I have so many more use cases I can work on now.

One such use case is attempting to move away from stock Arlo app notifications and create rich HTML5 notifications from HASS. I have this (mostly) working - I can now receive a push notification when motion is detected on a specific camera, along with the latest snapshot image (hooray). Here is the general flow of the automation:

Motion Detected -> Request Arlo Snapshot -> Camera Snapshot (to save image to file) -> Send Notification

The "Request Arlo Snapshot" is async, so I need to trigger the "Camera Snapshot" on a status change **from** 'taking snapshot' in order to be sure I get an updated image. 

**Problem:**  When the Arlo camera detects a motion event and starts its normal recording, the HASS entity status goes from "Idle" to "Recording" (briefly), then it quickly switches to "Taking snapshot" (from my automation firing), but then it STAYS on 'Taking snapshot' until the camera finishes its recording, after which it changes to "Recently Active". It does not switch from "Taking snapshot" back to "Recording" once the snapshot is done. Because of this, let's say a motion event is going on for a minute and a half - I won't get a notification for a minute and a half when the entity status changes from "Taking snapshot" to "Recently Active".  Is there a way to change the entity state to the last active state when the snapshot is actually finished creating instead of having to wait until a simultaneous recording finishes?

**Comment by twrecked** [comment]
Firstly, cool idea!

Right now the camera changes state based on messages received from Arlo. When you trigger a snapshot it moves into `fullFrameSnapshot` and then back into `Idle` and this is reflected on the card. If you're recording and trigger a snapshot I'm guessing it moves from `alertStreamActive` into `fullFrameSnapshot` and it never leaves that state until the recording finished. I say guessing because I've never tried it.

Having said that, when the snapshot finishes I do receive a `fullFrameSnapShotAvailable` so I can probably use that to return to the previous state and it'll trigger your notification quicker. Give me a couple of days to try it out.

**Comment by jwillaz** [comment]
Looking at the Arlo library, it appears there are two distinct snapshot endpoints. The one you are currently using (users/devices/fullFrameSnapshot) is for taking snapshots when the camera is _idle_. If the camera is _recording/streaming_, then snapshots have to be called via (users/devices/takeSnapshot).

That endpoint returns a `mediaUploadNotification` instead of `fullFrameSnapshotAvailable` event. The image is made available via `presignedContentUrl` instead of `presignedFullFrameSnapshotUrl`.

I would give it a try, but my Python/HASS knowledge is below mediocre, so I'm not quite sure how to migrate it to your refactored code.

**Comment by twrecked** [comment]
I can add that in, I'll probably create a new function for it and have the snapshot service call do the right thing depending on the camera state. I'll make the state transitions look like:

```
  idle --> takingSnapshot --> idle
  recording --> takingSnapshot --> recording
```

**Comment by twrecked** [comment]
Can you try this branch: https://github.com/twrecked/hass-aarlo/tree/stream-snapshot

**Comment by jwillaz** [comment]
I don't have access to test motion triggered recording, but my tests with triggered snapshot while streaming worked great! I did not notice any state change when taking the snapshot (it stayed on 'Streaming'), but the snapshot came through in the notification just fine and right after the expected save time. Thanks!

**Comment by twrecked** [comment]
Glad it worked. I think the state change is just too fast for the glance card - I see the Taking Snapshot message but it's gone pretty quickly. If you want to make sure you can look in the debug, you should see some messages like this: 

```
2019-03-12 17:25:20 DEBUG (MainThread) [pyaarlo] set:XXXXXXXXXX/activityState=fullFrameSnapshot
   ...a second or two later...
2019-03-12 17:21:53 DEBUG (ArloBackgroundWorker) [pyaarlo] our snapshot finished, restoring state
2019-03-12 17:21:53 DEBUG (ArloBackgroundWorker) [pyaarlo] set:XXXXXXXXXX/activityState=alertStreamActive
```

**Comment by twrecked** [comment]
Implemented in #31
